SPEC.md

```diff
@@ -310,6 +311,30 @@ Optional environment parameters:
 
 All parameters, with the exception of `CNI_PATH`, must be the same as the corresponding `ADD` for this container.
 
+
+#### `STATUS`: Check plugin status
+`STATUS` is a way for a runtime to determine the readiness of a network plugin.
+
+A plugin must exit with a zero (success) return code if the plugin is ready to service ADD requests. If the plugin is not able to service ADD requests, it must exit with a non-zero return code and output an error on standard out (see below).
+
+The following error codes are defined in the context of `STATUS`:
+
+- 50: The plugin is not available (i.e. cannot service `ADD` requests)
+- 51: The plugin is not available, and existing containers in the network may have limited connectivity.
+
+Plugin considerations:
+- Status is purely informational. A plugin MUST NOT rely on `STATUS` being called.
```

Comment:
That kills the install/upgrade use case; if you don't know for sure that the runtime is going to call `STATUS`, then you can't reliably use `STATUS` to signal plugin unavailability, so you have to fall back to the existing mechanism of not installing your CNI config until after the plugin is fully available.

Reply:
There are two issues here -- determining runtime CNI version and readiness racing.

CNI doesn't currently have a notion of, say, node readiness -- that is a Kubernetes concept. This PR provides a signal in the case where the runtime and plugins both care about readiness. However, there could always be runtimes (e.g. Podman) that don't have the notion of readiness, or even a node at all. The spec needs to support that use case.

I also want to strongly discourage STATUS being used as a pseudo-INIT.

That said, plugins that are operating in the broader kubernetes context do need a way to determine which behavior the runtime expects: status via file presence or status via STATUS. My proposal for that was a feature discovery mechanism (#927), and this is a good motivation to finish that proposal as well.

Reply:
> if you don't know for sure that the runtime is going to call STATUS, 

we need to provide the mechanisms to solve the problem,  and then take this changes to the container runtimes, as Dan says this does not work if the runtimes does not use it, but the plan is to get this into containerd and crio ...

Reply:
Follow-up: we couldn't find a great solution to this problem; plan is to release a small utility library that allows CNI-generating libraries to determine CNI version from the runtime version. This information is released via the kubernetes Node object, for example.